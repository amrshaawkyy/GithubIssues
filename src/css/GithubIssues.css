.navbar-center {
    display: flex;
    align-items: center;
    flex-direction: row;
    flex-wrap: wrap;
}
.navbar-center >p {
    margin-left: 15px;
}

/* .navbar.is-fixed-bottom, .navbar.is-fixed-top {
    width: 100%;
	position: fixed;
	bottom: 0;
	z-index: 1;
    background-color: red;
} */

.rotate {
    -moz-transition: all 2s linear;
    -webkit-transition: all 2s linear;
    transition: all 2s linear;
}

.DateRangePickerInput {
    border-radius: 10px;
    width: 70%;
    cursor: pointer;
    height: 40px;
    text-align: center;
    border: 1px solid lightgray;
    font-weight: bold;
    color: dimgrey;
}
.DateRangePickerInput:hover {
    background-color: lightgray;
}

.rotate.down {
    -moz-transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    transform: rotate(90deg);
}

.panel-icon {
    line-height: 0.5rem;
    text-align: center;
    vertical-align: middle;
    font-size: 10px;
    padding-top: 10%;
    padding-left: 10%;
}

/* .pagination{
    background: transparent;
} */

.pagination>button, button.pagination-link {
    background-color: #ffffff;
    border-radius: 50%;
}

/* .custom-section-1 {
    padding-top: 12px !important;
    padding-right: 0.75em;
    padding-bottom: 0.5em;
    padding-left: 0.75em;
} */

/* 
.fa-check {
    font-size: 100%;
    margin: -0.01rem 0 0 0.9rem;
} */

.tag {
    margin: 2px;
    height: 1.2em !important;
}

.time {
    font-size: 0.8rem;
    width: 100%;
}

.bodyContainer {
    max-height: fit-content;
    display: block;
    background-color: #ffffff;
    margin-top: 10px;
    overflow-wrap: normal;
    overflow: hidden;
    padding: 10px !important;
}

.bodyContainer>li, ol, pre, textarea, ul {
    padding: inherit;
    /* margin: 10px !important; */
}

/* .panel-heading {
    padding-top: 0.5em;
    padding-right: 0.75em;
    padding-bottom: 0em;
    padding-left: 0.75em;
} */

.panel-block, .panel-heading, .panel-tabs {
    border-top: 1px solid #dbdbdb;
    border-bottom: 1px solid #dbdbdb;
    border-left: 1px solid #dbdbdb;
    border-right: 1px solid #dbdbdb;
}

.panel-block {
    flex-flow: wrap;
    align-items: center;
    color: #363636;
    padding: .5em .75em;
    width: 100%;
    margin: 0px !important;
    cursor: pointer;
}

.panel-ovr {
    border: none !important;
    margin: 0px !important;
    width: 100%;
    max-width: 100%;
    line-height: 1.25;
}

._title {
    text-overflow: ellipsis;
    overflow-x: hidden;
    max-width: 90%;
    overflow-wrap: normal;
    color: #363636 !important;
    margin-right: 12px;
    font-weight: 900;
}

.milestone, .comments {
    margin-left: 20px;
}

.show-body {
    padding: 5px;
}

.panel-icon-container {
    margin-right: -3.5% !important;
}

.info-container {
    margin-left: -2vw;
    margin-top: 2vw;
    /*check*/
    line-height: 1;
}

@media screen and (max-width: 769px) {
    .panel-icon-container {
        margin-right: 0 !important;
    }
    .info-container {
        margin-left: 0% !important;
        padding-left: 40px;
    }
    .milestone {
        margin-left: 0px !important;
    }
    .title-string {
        font-size: 4vw;
    }
}