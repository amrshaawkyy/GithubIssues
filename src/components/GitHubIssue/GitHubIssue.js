import React from 'react';
import WithErrorHandler from '../../containers/withErrorHandler/WithErrorHandler';
import Aux from "../../hoc/AUX1/Aux1";
import Icons from '../UI/Icons/Icons';
import CreationTime from '../Information/CreationTime';
import Labels from '../Information/Labels';
import { faThumbtack } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Comments from '../Information/Comments';
import PropTypes from 'prop-types';
import '../../css/GithubIssues.css';
const showdown = require('showdown')
const gitHubIssue = (props) => {
    if (props.state.error) {
        return <WithErrorHandler errormsg={props.state.error.message + ": Can Not Fetch Data"} />;
    }
    const isVisible = (id) => {
        return props.state.showBody["_" + id] ? " container bodyContainer" : " is-hidden"
    }
    const handleClick = (e) => {
        e.preventDefault();
        let converter = new showdown.Converter({
            tables: true,
            strikethrough: true,
            ghCompatibleHeaderId: true,
            literalMidWordUnderscores: true,
            ghCodeBlocks: true,
            tasklists: true,
            ghMentions: true,
            ghMentionsLink: 'https://github.com/{u}'
        })
        let obj = {}
        const id = e.currentTarget.dataset.id;
        obj["_" + id] = props.state.showBody["_" + id] === true ? false : true
        props.showBodyHandler(obj)
        if (obj["_" + id]) {
            props.state.issues.map(function (issues) {
                if (issues.id === +id) {
                    return document.querySelector(`._${id}`).innerHTML = converter.makeHtml(issues.body)
                }
                return true
            })
        }
    }
    const hideMe = (issueOrPr) => {
        let choiceType = issueOrPr ? "issues" : "pull requests"
        return (props.state.listFilter.choice !== (choiceType || "all") ? " " : " is-hidden")
    }
    return (
        < Aux >
            {
                props.state.issues.map(function (data, index) {
                    return <a href="/" key={index} className={"panel-block" + hideMe(data.pull_request)}>
                        <span data-id={data.id} className="panel-ovr" onClick={handleClick} >
                            <span className="columns is-mobile is-multiline is-vcentered is-2">
                                <span className=" column is-1">
                                    <span className="panel-icon icon is-small">
                                        <Icons type={!data.pull_request} state={data.state} />
                                    </span>
                                </span>
                                <span className="column is-11 info-container">
                                    <span className="_title">{data.title}</span>
                                    <br className="is-hidden-tablet" />
                                    <span className="time is-hidden-tablet">
                                        <CreationTime data={data} />
                                    </span>
                                    <br className="is-hidden-tablet" />
                                    <Labels labels={data.labels} />
                                    <br className="is-hidden-mobile" />
                                    <span className="time is-hidden-mobile">
                                        <CreationTime data={data} />
                                        <Comments data={data} />
                                    </span>
                                </span>
                                <span className="column show-body has-text-right is-pulled-right">
                                    <FontAwesomeIcon /*className="rotate"*/ icon={faThumbtack} style={{ color: 'blue' }} />
                                </span>
                            </span>
                        </span>
                        <span className={"_" + data.id + (isVisible(data.id))}></span>
                    </a>;
                })
            }
        </ Aux>
    );
}
gitHubIssue.propTypes = {
    error: PropTypes.string,
    showBody: PropTypes.string,
    issues: PropTypes.array
};
export default gitHubIssue;