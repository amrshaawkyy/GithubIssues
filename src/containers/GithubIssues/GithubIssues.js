import React, { Component } from 'react';
import IssuesFiltering from '../IssuesFiltering/IssuesFiltering';
import Pagination from '../../components/UI/Pagination/Pagination';
import 'bulma/css/bulma.css';
import '../../css/GithubIssues.css';
import Spinner from '../../components/UI/Spinner/Spinner';
import GitHubIssue from '../../components/GitHubIssue/GitHubIssue';
class GithubIssues extends Component {
    state = {
        user: 'facebook',
        repo: 'react',
        listFilter: { state: "open", choice: "issues" },
        since: new Date('01/01/'+ new Date().getFullYear()),
        per_page: 40,
        sort: true,
        page: '',
        currentPageNumber: 1,
        pages: {},
        issues: [],
        loading: true,
        error: null,
        showBody: {},
        sortName: "Latest"
    }
    componentDidMount() {
        this.fetchAllIssues();
    }
    shouldComponentUpdate(previousState, nextState) {
        if (previousState.page !== nextState.page) {
            return true;
        }
        if (previousState.listFilter !== nextState.listFilter) {
            return true;
        }
        return false;
    }
    fetchAllIssues() {
        const headers = {
            headers: {
                Authorization: "token 60e07ff280c786e523a632be9af8f992270a5c5b",
                Accept: "application/vnd.github.v3+json,application/vnd.github.machine-man-preview+json",
            }
        };
        const baseUrl = "https://api.github.com/repos"
        let url = { state: this.state.listFilter.state, since: this.state.since, per_page: 40, sort: this.state.sort ? "created" : "created-asc" };
        const keys = Object.keys(url)
        const params = keys.length ?
            "?" + keys
                .map(key => encodeURIComponent(key) +
                    "=" + encodeURIComponent(url[key]))
                .join("&") :
            "";
        let linkHeaders = ''
        let userRepoIssues = `${this.state.user}/${this.state.repo}/issues`
        let fullUrl = `${baseUrl}/${userRepoIssues}${params}${this.state.page}`
        fetch(fullUrl, headers)
            .then(response => {
                if (response.headers.get('Link')) {
                    linkHeaders = this.parseLinkHeader(response.headers.get('Link'))
                }
                if (response.ok) return response.json();
                throw new Error('Request failed');
            })
            .then(data => {
                this.setState({
                    pages: linkHeaders,
                    issues: data,
                    loading: false,
                    error: null,
                    showBody: {},
                });
            })
            .catch(error => {
                this.setState({
                    loading: false,
                    error: error
                });
            });
    }
    handlePageChange(page) {
        const newPage = (previousState) => {
            return { ...previousState, page: page, currentPageNumber: +page.split('=').pop(), loading: true };
        };
        this.setState(newPage, () => {
            this.fetchAllIssues()
        })
    }
    handleFilterChange(toFilter) {
        let newObj = { ...this.state.listFilter, ...toFilter }
        const newFilter = (previousState) => {
            let stateTest = previousState.listFilter.state !== newObj.state ? true : false
            return { ...previousState, listFilter: newObj, loading: stateTest };
        };
        this.setState(newFilter, () => {
            Object.keys(toFilter)[0] === "state" && this.fetchAllIssues()
        })
    }
    parseLinkHeader = (header) => {
        if (header.length === 0) {
            throw new Error("input must not be of zero length");
        }
        return header.split(/(?!\B"[^"]*),(?![^"]*"\B)/).reduce((links, part) => {
            const section = part.split(/(?!\B"[^"]*);(?![^"]*"\B)/);
            if (section.length < 2) {
                throw new Error("section could not be split on ';'");
            }
            const url = section[0].replace(/<(.*)>/, '$1').trim();
            const name = section[1].replace(/rel="(.*)"/, '$1').trim();
            links[name] = url;

            return links;
        }, {});
    }
    handleSortingChange() {
        let newObj = { ...this.state.sort };
        const newState = (previousState) => {
            let stateTest = previousState.sort !== newObj ? true : false
            return { ...previousState, sort: !previousState.sort, loading: stateTest, sortName: !previousState.sort ? "Latest" : "Oldest" };
        };
        this.setState(newState, () => {
            this.fetchAllIssues()
        })
    }
    showBodyHandler = (obj) => {
        this.setState({ showBody: obj })
    }
    handleDateChange (date)  {
        const newFilter = (previousState) => {
            let stateTest = previousState.since !== date ? true : false
            return { ...previousState, since: date, loading: stateTest };
        };
        this.setState(newFilter, () => {
            this.fetchAllIssues()
        })
    }
    render() {
        return (
            <div className="container">
                <nav className="panel" >
                    <IssuesFiltering
                        listFilter={this.state.listFilter}
                        onFilterChange={(event) => this.handleFilterChange(event)}
                        sortName={this.state.sortName}
                        clicked={(event) => this.handleSortingChange(event)}
                        startDate={this.state.since}
                        handleDateChange={(event)=>this.handleDateChange(event)}
                    />
                    {this.state.loading ?
                        <Spinner />
                        : <GitHubIssue state={this.state} showBodyHandler={this.showBodyHandler} />}
                    <nav className="navbar  is-fixed-bottom is-transparent">
                        <div className="navbar-start" style={{ flexGrow: 1, justifyContent: "center" }}>
                            <div className="navbar-item">
                                <Pagination pages={this.state.pages} currentPageNumber={this.state.currentPageNumber} onPageChange={(event) => this.handlePageChange(event)} />
                            </div>
                        </div>
                    </nav>
                </nav></div>);
    }
}
export default GithubIssues;

