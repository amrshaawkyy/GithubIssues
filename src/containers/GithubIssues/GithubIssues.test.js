import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import  GithubIssues  from './GithubIssues';
import GitHubIssue from '../../components/GitHubIssue/GitHubIssue';
import Spinner from '../../components/UI/Spinner/Spinner';

configure({adapter: new Adapter()});

describe('<GithubIssues />', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<GithubIssues />);
    });
    afterEach(() => {
        expect(wrapper).toMatchSnapshot();
    });
    it('should render <GitHubIssue /> when Loading is false', () => {
        wrapper.setState({loading: false});
        expect(wrapper.find(GitHubIssue).length).toBe(1);
    });
    it('should render <Spinner /> when Loading is True', () => {
        wrapper.setState({loading: true});
        expect(wrapper.find(Spinner).length).toBe(1);
    });
});


