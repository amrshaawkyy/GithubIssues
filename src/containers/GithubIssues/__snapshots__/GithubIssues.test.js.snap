// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GithubIssues /> should render <GitHubIssue /> when Loading is false 1`] = `ShallowWrapper {}`;

exports[`<GithubIssues /> should render <Spinner /> when Loading is True 1`] = `ShallowWrapper {}`;
